MudTabs: Fix JsInteropException when trimmed binary (#6050)

Swizzy opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
<!-- Describe your changes in detail any why -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310 -->

Fixes #6050

I implemented a CustomJson converter for BoundingClientRect to resolve a issue i encountered when using the latest build of MudBlazor, during investigation of the issue i found out that the issue only happened in trimmed binaries.

The converter only supports deserialization as i could not find any use case for it being needed for serialization.

## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

It's both been tested using my own Blazor application with a custom build of MudBlazor and using a Unit test to verify that it doesn't crash and gets the correct values as expected.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

Swizzy committed:
MudTabs: Fix JsInteropException when trimmed binary (#6050)

mikes-gh commented:
@ScarletKuro Would you mind reviewing please.
Why do the Docs work?
Is it due to this
https://github.com/MudBlazor/MudBlazor/blob/dc399ae15b7cff79d6859d337beab78d445169cb/src/MudBlazor.Docs.Wasm/MudBlazor.Docs.Wasm.csproj#L9

mikes-gh commented:
@Swizzy Just checking.  Are you using AOT?

Swizzy commented:
> @Swizzy Just checking. Are you using AOT?

No, i don't use anything special, i just set this project up the other day from one of the .NET 7 empty templates, added MudBlazor and JSON.NET to resolve another serialization issue where i am in more control.

mikes-gh commented:
Ok so this could be a JSON.net issue?
I really need a repro before merging this since the Docs is working.

ScarletKuro commented:
@mikes-gh 
> Is it due to this <_AggressiveAttributeTrimming>false</_AggressiveAttributeTrimming>?

No, this just prevents from trimming some attributes that Docs is using, for example `System.ObsoleteAttribute`.

This is most likely because of this line
https://github.com/MudBlazor/MudBlazor/blob/5d6ba1b42ab3a444fe68e0329ee8b44533171aba/src/MudBlazor/Services/ReseizeObserver/ResizeObserver.cs#L58

Even thought `InvokeAsync`(it's converting the object into json inside) is annotated `InvokeAsync<[DynamicallyAccessedMembers(JsonSerialized)]` but this is wrapped into an `IEnumerable` so I guess some properties in`BoundingClientRect` get trimmed. I do not think this requires adding a JsonConverter and you can fix it with `DynamicDependency`.

mikes-gh commented:
OK so we still need to answer the question why does it work on Docs?

BTW @Swizzy I think you're going to have issues with JSON.net with trimming.  `System.Text.Json` has good trimming support including code generators which NewtonSoft have said they wont support as its too much work.

ScarletKuro commented:
Maybe because IconsPage is referencing BoundingClientRect? It's theory only

ScarletKuro commented:
Is this a reproduction repo https://github.com/Swizzy/ZENBitPackToolbox ?

Swizzy commented:
> Is this a reproduction repo https://github.com/Swizzy/ZENBitPackToolbox ?

I had the issue there before my last commit yes - i'm trying to set up a cleaner repo for you guys with nothing but this issue to make it easier to narrow things down :)

Swizzy commented:
A repro repo can now be found here; https://github.com/Swizzy/MudBlazorJSInteropRepro

As minimalist as possible; setup a new template, added MudBlazor (and only MudBlazor), made the changes as per the instructions to include the CSS/JS and added a MudTabs component with 3 pages with no special content and made it deploy to GH Pages (where i personally found the issue).

you can see it in action here; https://swizzy.github.io/MudBlazorJSInteropRepro

Swizzy commented:
> OK so we still need to answer the question why does it work on Docs?
> 
> BTW @Swizzy I think you're going to have issues with JSON.net with trimming. `System.Text.Json` has good trimming support including code generators which NewtonSoft have said they wont support as its too much work.

Thanks for the heads up, i'll look into making it work with System.Text.Json, but i felt like switching over to NewtonSoft to be an OK quick fix as it solved the problem i had at the time (not related to this one)

Swizzy commented:
> @mikes-gh
> 
> > Is it due to this <_AggressiveAttributeTrimming>false</_AggressiveAttributeTrimming>?
> 
> No, this just prevents from trimming some attributes that Docs is using, for example `System.ObsoleteAttribute`.
> 
> This is most likely because of this line
> 
> https://github.com/MudBlazor/MudBlazor/blob/5d6ba1b42ab3a444fe68e0329ee8b44533171aba/src/MudBlazor/Services/ReseizeObserver/ResizeObserver.cs#L58
> 
> Even thought `InvokeAsync`(it's converting the object into json inside) is annotated `InvokeAsync<[DynamicallyAccessedMembers(JsonSerialized)]` but this is wrapped into an `IEnumerable` so I guess some properties in`BoundingClientRect` get trimmed. I do not think this requires adding a JsonConverter and you can fix it with `DynamicDependency` -> `[DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(BoundingClientRect))]` to the `ResizeObserver`.

I tried adding `[DynamicallyAccessedMembers(PublicConstructors)]` and later also `[DynamicallyAccessedMembers(PublicParameterlessConstructor)]` with and without a empty parameterless constructor without success previously.

My first thought was that that should be sufficient aswell, it's entirely likely that what you mentioned is the reason the docs work - if those somewhere instantiate a new BoundingClientRect that could very well be the reason it works fine in the docs as it obviously then uses the constructor that is missing according to the error messages

ScarletKuro commented:
Ty for repository, I could reproduce it. It indeed gets trimmed, if you open with DotPeetk for example, thats what you will see inside the `BoundingClientRect`

```CSharp
namespace MudBlazor.Interop
{
  public class BoundingClientRect
  {
    public double Width
    {
      [MethodImpl(MethodImplOptions.NoInlining)] get => throw new NotSupportedException("Linked away");
    }

    public double Height
    {
      [MethodImpl(MethodImplOptions.NoInlining)] get => throw new NotSupportedException("Linked away");
    }
  }
}
```

mikes-gh commented:
I also reproed it here https://mudblazor-test2.azurewebsites.net/ in azure.

ScarletKuro commented:
`[DynamicallyAccessedMembers(PublicParameterlessConstructor)]` are not enough, because the properties gets trimmed as well, they are needed for JS interop as well. I verified that `[DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(BoundingClientRect))]`. Instead of All we probably could use `DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.PublicProperties | DynamicallyAccessedMemberTypes.NonPublicConstructors` but since we started to use `All` I think it's better to have the code consistency.

ScarletKuro commented:
I also noticed that linker from .net6 wasn't touching this class, maybe this is why it was left unnoticed for this long. This kind of interop is really tricky, since the compiler won't give you any warrning.

mikes-gh commented:
> `[DynamicallyAccessedMembers(PublicParameterlessConstructor)]` is not enough, because the properties gets trimmed as well, they are needed for JS interop as well. I verified that `[DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(BoundingClientRect))]` works. Instead of All we probably could use `DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.PublicProperties | DynamicallyAccessedMemberTypes.NonPublicConstructors` but since we started to use `All` I think it's better to leave as as to have the code consistency.

Thanks @ScarletKuro I think we should use the minimum required access but thats me.  When there is lots of member types I guess the trimming difference is less.  But I think `All` was used initially when we were discovering triming and we shouldnt use it as the standard

ScarletKuro commented:
>  I think All was used initially when we were discovering triming

Yeah. And you are right, that we should use minimum required for optimization, but maybe we should do it as separate PR? Up to you, idm any variant.

mikes-gh commented:
Closed in favour of https://github.com/MudBlazor/MudBlazor/pull/6053
@Swizzy Thanks for the report

mikes-gh commented:
> 



> > I think All was used initially when we were discovering triming
> 
> Yeah. And you are right, that we should use minimum required for optimization, but maybe we should do it as separate PR? Up to you, idm any variant.

We can narrow down access in another PR

ScarletKuro commented:
Ok. I left comment in the PR for future.
